src/cryptography/hazmat/backends/openssl/dh.py

```diff
@@ -8,14 +8,24 @@
 from cryptography.hazmat.primitives.asymmetric import dh
 
 
-def _dh_cdata_to_parameters(dh_cdata, backend):
+def _dh_params_dup(dh_cdata, backend):
     lib = backend._lib
     ffi = backend._ffi
 
     param_cdata = lib.DHparams_dup(dh_cdata)
     backend.openssl_assert(param_cdata != ffi.NULL)
     param_cdata = ffi.gc(param_cdata, lib.DH_free)
+    if lib.OPENSSL_VERSION_NUMBER < 0x10002000:
+        # In OpenSSL versions < 1.0.2 DHparams_dup don't copy q
+        q = ffi.new("BIGNUM **")
+        lib.DH_get0_pqg(dh_cdata, ffi.NULL, q, ffi.NULL)
```

Comment:
the returned bignum is not retained (`get0`) so you need to do a BN_dup here:

`q_dup = lib.BN_dup(q[0])`

`BN_dup` will return null if passed null so this is safe to do without an assertion. Then you can pass `q_dup` to `DH_get0_pqg`.